[EDR Workflows][Fleet] Search and display policy name for uninstall tokens

gergoabraham opened the pull request:
## Summary

In Fleet / Uninstall Tokens tab:
- user can see the name of the related Agent Policy in a new column, if the policy still exists,
<img width="1325" alt="image" src="https://github.com/elastic/kibana/assets/39014407/1b06e39b-103f-4439-9560-51227d75fb25">


- otherwise an info tooltip indicating why the name is missing.
test: `This token's related Agent policy has already been deleted, so the policy name is unavailable.`
<img width="1328" alt="image" src="https://github.com/elastic/kibana/assets/39014407/d98a72d7-9563-4d4e-9390-9d9826702026">



- Policy name (and the tooltip if missing) is indicated in Uninstall Command flyout, as well.
<img width="300" alt="image" src="https://github.com/elastic/kibana/assets/39014407/38d33f3e-5f51-4c88-9efc-db9aed10ad42"><img width="300" alt="image" src="https://github.com/elastic/kibana/assets/39014407/d1da0177-5e21-44cb-9ef7-e768d02a00aa">

- User can search for Uninstall Tokens by policy ID and policy name as well: the search is a **case sensitive partial match**, excluding any special character
(note: the text on the top of each screenshot has been updated, see first screenshot)

![image](https://github.com/elastic/kibana/assets/39014407/a5461186-a063-4dc3-8f50-3a6fd0af626d)

![image](https://github.com/elastic/kibana/assets/39014407/2ccaffd7-1358-4113-8e4b-a930d5e20d0f)

![image](https://github.com/elastic/kibana/assets/39014407/d6840f08-ac4f-4d72-b827-abbff25c146a)

![image](https://github.com/elastic/kibana/assets/39014407/c18f4437-1fc9-4c4e-b846-21b68651ebb9)


- a hint is added to indicate that a deleted agent policy's name is unknown
test: `If an Agent policy is deleted, its policy name is also deleted. Use the policy ID to search for uninstall tokens related to deleted Agent policies.`
<img width="1308" alt="image" src="https://github.com/elastic/kibana/assets/39014407/816dae7e-e3d3-445f-8ee1-48f93bd4f69a">




### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

elasticmachine commented:
Pinging @elastic/security-defend-workflows (Team:Defend Workflows)

gergoabraham commented:
hey @roxana-gheorghe, @joepeeples (or @elastic/security-docs), could you please check the texts? I've copied those to the PR description, and they're also on screenshots

thank you!

juliaElastic reviewed (approved):
Fleet changes LGTM

gergoabraham commented:
@elasticmachine merge upstream

ashokaditya reviewed (approved):
Love the names on the screenshot. 😄  I've a few suggestion but not blockers for 🚢 ing this PR.

gergoabraham commented on x-pack/plugins/fleet/public/components/uninstall_command_flyout/empty_policy_name_hint.tsx:
```diff
@@ -0,0 +1,28 @@
+  }
+);
+
+export const EmptyPolicyNameHint = () => (
```
this looks good, Ash! thanks for the suggestion, in case there will be no objection soon, I'll include it

<img width="622" alt="image" src="https://github.com/elastic/kibana/assets/39014407/08e0d319-4c86-4219-adc4-9d11589b4d7b">

<img width="472" alt="image" src="https://github.com/elastic/kibana/assets/39014407/a6ad36df-d36f-4fbd-b220-e86e34d8cc5f">

ashokaditya replied:
Thanks @gergoabraham . Yeah the row with just an icon looks very different from the other rows and a user might miss it thinking it is just a grouping/divider icon. I'm hoping with a `-`. it is slightly more visible. I think we could also show a `No name available` in italics or a subdued colour maybe be even better.

gergoabraham replied:
added here: 703c519e73bbebcd2c622c555214fd797786f83c

gergoabraham replied:
I think once the user reads the hint next to the dash, it will be clear to them what the dash means (probably even before that), so it's better not pollute the screen with 'No name available' strings, there could be a lot of them which can be distractive. I'd go with the dash - what do you think?

ashokaditya replied:
Yeah either is fine with me.

gergoabraham commented on x-pack/plugins/fleet/server/routes/uninstall_token/handlers.ts:
```diff
@@ -44,10 +54,12 @@ export const getUninstallTokensMetadataHandler: FleetRequestHandler<
     const managedPolicyIds = managedPolicies.map((policy) => policy.id);
 
-    const { page = 1, perPage = 20, policyId } = request.query;
+    const policyIdSearchTerm = search?.trim() ?? policyId?.trim();
```
yeah, that's shady, thanks for pointing that out! tried to refactor it, so maybe now it reflects the logic better. please let me know if it helps: 0d410b709562df40abb38fb64c130b9ffb4ab541
(if not, I can add the comment, but as long as it's not a must, I don't want to ^^)

gergoabraham replied:
one additional thing: here it's not about the deleted policy, but about using either the original `policyId` query param to filter by policy ID only, or the new `search` query param which filters both policy ID and name the same time.

but again, this just proves that the logic is, or hopefully was, shady

gergoabraham commented on x-pack/plugins/fleet/cypress/e2e/uninstall_token.cy.ts:
```diff
@@ -16,18 +16,25 @@ import { request } from '../tasks/common';
+        generatePolicies();
+
+        if (removePolicies) {
+          cleanupAgentPolicies();
```
here we have 2 test groups, in one we remove the agent policies, in the other they still exist, before checking uninstall tokens, see the `forEach` in line 19

so when `removePolicies` is true, in this `before` hook we first do some initial cleanup, then generate new agent policies, then delete those newly generated policies while their uninstall tokens keep existing

ashokaditya replied:
Oh indeed. You're cleaning up the generated policies. I missed that.

ashokaditya replied:
Oh indeed. You're cleaning up the generated policies. I missed that.

joepeeples commented:
@gergoabraham Thanks for pinging for UI text review — love your policy names in the screenshots. 😋 Some UI text edit suggestions:

**Page description:**
_(Not directly related to the policy name, but this is a good opportunity to tweak this description)_

An uninstall token allows you to use the uninstall command to remove Elastic Agent from a host.

**Tooltip - missing name:**
This token’s related Agent policy has already been deleted, so the policy name is unavailable.

**Search tooltip “hint”:**
If an Agent policy is deleted, its policy name is also deleted. Use the policy ID to search for uninstall tokens related to deleted Agent policies.

gergoabraham commented:
@joepeeples awesome, thanks for the suggestions! somehow I felt that `hence` needs to go. 😀

updated here: 2c7244a5a1f821b2bf0da64818ad9d50cb3f9f67
and also updated the screenshots/text in the PR description

ashokaditya reviewed (approved):
Thanks for the changes.
